File: cpp/medium/SubarraySum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

int subarraySum(vector<int> &nums, int k) {
  int rs = (nums[0] == k ? 1 : 0);
  unordered_map<int, int> map;
  for (int i = 1; i < nums.size(); ++i) {
    nums[i] += nums[i - 1];
    if (nums[i] == k)
      rs++;
    if (map.count(nums[i] - k))
      rs += map[nums[i] - k];
    if (map.count(nums[i]))
      map[nums[i]]++;
    else
      map[nums[i]] = 1;
  }
  return rs;
}

int main() {
  vector<int> nums = {1, 1, 1};
  cout << subarraySum(nums, 2);
  return 0;
}

File: rust/src/easy/subarray_sum.rs
use std::cmp::max;
use std::collections::hash_set::Union;
use std::path::absolute;

pub fn subarray_sum(nums: Vec<i32>) -> i32 {
    let mut prefix_sum = vec![0; nums.len()];
    prefix_sum[0] = nums[0];
    let mut sum = prefix_sum[0];
    for i in 1..nums.len() {
        prefix_sum[i] += prefix_sum[i - 1] + nums[i];
        sum += prefix_sum[i] - prefix_sum[max(0,i as i32 - nums[i]) as usize]
    }
    sum
}